server/path_watcher.go

```diff
@@ -0,0 +1,309 @@
+
+	ctx, cancel := context.WithCancel(parentCtx)
+	defer cancel()
+
+	watcher.previousProcessedRevision = watcher.fetchStoredRevision(parentCtx)
+	eventsCh := watcher.sync.Watch(ctx, watcher.path, watcher.previousProcessedRevision+1)
+	doneCh := make(chan error, 1)
+
+	go watcher.consumeEvents(ctx, eventsCh, doneCh)
+
+	select {
+	case <-ctx.Done():
+		if err := <-doneCh; err != nil {
+			log.Error("%s consuming events failed: %s", watcher, err)
+		}
+		return ctx.Err()
+	case <-lost:
+		watcher.updateCounter(1, "lock.lost")
+		cancel()
+		if err := <-doneCh; err != nil {
+			log.Error("%s error after lost lock: %s", watcher, err)
+		}
+		return fmt.Errorf("lock for path `%s` is lost", watcher.path)
+	case err := <-doneCh:
+		return err
+	}
+}
+
+func (watcher *PathWatcher) consumeEvents(ctx context.Context, eventCh <-chan *gohan_sync.Event, watchErr chan<- error) {
+	watcher.updateCounter(1, "running")
+	defer watcher.updateCounter(-1, "running")
+
+	var err error
+	defer func() {
+		watchErr <- err
+	}()
+
+	for {
+		select {
+		case event, ok := <-eventCh:
+			if !ok {
+				err = ctx.Err()
+				return
+			}
+			if err = watcher.consumeEvent(event); err != nil {
+				return
+			}
+		case <-ctx.Done():
+			err = ctx.Err()
+			return
+		}
+	}
+}
+
+func (watcher *PathWatcher) consumeEvent(event *gohan_sync.Event) (err error) {
+	// each event & recovery must not be interrupted,
+	// otherwise we could easily end up in an inconsistent state
+	// if ctx is canceled either in running extensions or just before recovery
+	ctx := context.Background()
+
+	defer func() {
+		watcher.tryRecover(ctx, err, event)
+	}()
+
+	if event.Err != nil {
+		return event.Err
+	}
+	watcher.watchExtensionHandler(ctx, event)
+
+	return watcher.storeRevision(ctx, event.Revision)
+}
+
+func (watcher *PathWatcher) tryRecover(ctx context.Context, err error, event *gohan_sync.Event) {
+	if err == errInconsistentCluster {
+		watcher.tryRecoverInconsistentCluster(ctx, event)
+	} else if errCompacted, ok := err.(goext.ErrCompacted); ok {
+		watcher.tryRecoverCompaction(ctx, event, errCompacted.CompactRevision)
+	}
+}
+
+func (watcher *PathWatcher) tryRecoverInconsistentCluster(ctx context.Context, event *gohan_sync.Event) {
+	node, err := watcher.sync.Fetch(ctx, event.Key)
+	if err != nil {
+		watcher.updateCounter(1, "inconsistency_recovery.fetch_errors")
+		log.Critical("%s can't recover from inconsistent cluster state on %s: Fetch() failed: %s", watcher, event.Key, err)
+		return
+	}
+
+	if recovered, err := watcher.sync.CompareAndSwap(ctx, event.Key, node.Value, watcher.sync.ByValue(node.Value)); err != nil {
+		// some sync events could have be processed out-of-order and the recovery failed.
+		// incorrect data could be stored in DB. a user has to recover (resync) manually
+		watcher.updateCounter(1, "inconsistency_recovery.cas_errors")
+		log.Critical("%s can't recover from inconsistent cluster state on %s: notifying the current master failed: %s", watcher, event.Key, err)
```

Comment:
Lots of very long lines until the end of this file.

Reply:
formatting of logs improved